neo/Network/P2P/Peer.cs

```diff
@@ -0,0 +1,219 @@
+        }
+
+        protected void AddPeers(IEnumerable<IPEndPoint> peers)
+        {
+            if (UnconnectedPeers.Count < UnconnectedMax)
+            {
+                peers = peers.Where(p => p.Port != ListenerPort || !localAddresses.Contains(p.Address));
+                ImmutableInterlocked.Update(ref UnconnectedPeers, p => p.Union(peers));
+            }
+        }
+
+        protected void ConnectToPeer(IPEndPoint endPoint)
+        {
+            endPoint = endPoint.Unmap();
+            if (endPoint.Port == ListenerPort && localAddresses.Contains(endPoint.Address)) return;
+            if (ConnectedAddresses.TryGetValue(endPoint.Address, out int count) && count >= MaxConnectionsPerAddress)
+                return;
+            if (ConnectedPeers.Values.Contains(endPoint)) return;
+            tcp_manager.Tell(new Tcp.Connect(endPoint));
+        }
+
+        private static bool IsIntranetAddress(IPAddress address)
+        {
+            byte[] data = address.MapToIPv4().GetAddressBytes();
+            Array.Reverse(data);
+            uint value = data.ToUInt32(0);
+            return (value & 0xff000000) == 0x0a000000 || (value & 0xff000000) == 0x7f000000 || (value & 0xfff00000) == 0xac100000 || (value & 0xffff0000) == 0xc0a80000 || (value & 0xffff0000) == 0xa9fe0000;
+        }
+
+        protected abstract void NeedMorePeers(int count);
+
+        protected override void OnReceive(object message)
+        {
+            switch (message)
+            {
+                case Start start:
+                    OnStart(start.Port, start.WsPort);
+                    break;
+                case Timer _:
+                    OnTimer();
+                    break;
+                case Peers peers:
+                    AddPeers(peers.EndPoints);
+                    break;
+                case Connect connect:
+                    ConnectToPeer(connect.EndPoint);
+                    break;
+                case WsConnected ws:
+                    OnWsConnected(ws.Socket, ws.Remote, ws.Local);
+                    break;
+                case Tcp.Connected connected:
+                    OnTcpConnected(((IPEndPoint)connected.RemoteAddress).Unmap(), ((IPEndPoint)connected.LocalAddress).Unmap());
+                    break;
+                case Tcp.Bound _:
+                    tcp_listener = Sender;
+                    break;
+                case Tcp.CommandFailed _:
+                    break;
+                case Terminated terminated:
+                    OnTerminated(terminated.ActorRef);
+                    break;
+            }
+        }
+
+        private void OnStart(int port, int ws_port)
+        {
+            ListenerPort = port;
+            timer = Context.System.Scheduler.ScheduleTellRepeatedlyCancelable(0, 5000, Context.Self, new Timer(), ActorRefs.NoSender);
```

Comment:
How about changing this to wait 500ms or so initially instead of 0ms, so that immediately after starting the node one can Tell LocalNode to connect to some previously connected nodes prior to going out and hitting seed servers.

Previously the code supported reading back peers from a file that is saved on shutdown. To continue supporting this behavior, I've added an initial delay and Tell Peer.Connect to previously connected nodes.

Reply:
You can `Tell` some `Peer.Connect` objects to it before `Tell`ing the `Start` object.

Reply:
Ah, I haven’t used Akka much, and wasn’t aware you could queue messages before starting. I’ll give that a test.

Reply:
Are you sure it is a good idea for it to be doing Peer.Connect before Start? This will make the Peer.Connect message get processed before Start and it will not yet be listening locally for TCP or websocket connections?

Also, suppose later more logic is added to start that is meant to happen before other messages are processed. I think this is a bad idea personally. Maybe the Start message could include a list of peers to connect to initially to support that use case?

Reply:
I saw that was possible; but I wanted to directly tell it to connect to them initially. I suppose if I’m adding them as you describe it should naturally connect to them though; so will try that.